scvi/models/_base.py

```diff
@@ -0,0 +1,792 @@
+            Minibatch size for data loading into model
+        return_mean
+            Whether to return the mean of the samples.
+        return_numpy
+            Return a `np.ndarray` instead of a `pd.DataFrame`. Includes gene
+            names as columns. If either n_samples=1 or return_mean=True, defaults to False.
+            Otherwise, it defaults to True.
+
+        Returns
+        -------
+        - **normalized_expression** - array of normalized expression
+
+        If ``n_samples`` > 1 and ``return_mean`` is False, then the shape is ``(samples, cells, genes)``.
+        Otherwise, shape is ``(cells, genes)``. Return type is ``pd.DataFrame`` unless ``return_numpy`` is True.
+        """
+        adata = self._validate_anndata(adata)
+        post = self._make_posterior(adata=adata, indices=indices, batch_size=batch_size)
+
+        if gene_list is None:
+            gene_mask = slice(None)
+        else:
+            all_genes = _get_var_names_from_setup_anndata(adata)
+            gene_mask = [True if gene in gene_list else False for gene in all_genes]
+
+        if n_samples > 1 and return_mean is False:
+            if return_numpy is False:
+                logger.warning(
+                    "return_numpy must be True if n_samples > 1 and return_mean is False, returning np.ndarray"
+                )
+            return_numpy = True
+        if indices is None:
+            indices = np.arange(adata.n_obs)
+        if library_size == "latent":
+            model_fn = self.model.get_sample_rate
+            scaling = 1
+        else:
+            model_fn = self.model.get_sample_scale
+            scaling = library_size
+
+        exprs = []
+        for tensors in post:
+            x = tensors[_CONSTANTS.X_KEY]
+            batch_idx = tensors[_CONSTANTS.BATCH_KEY]
+            labels = tensors[_CONSTANTS.LABELS_KEY]
+            exprs += [
+                np.array(
+                    (
+                        model_fn(
+                            x,
+                            batch_index=batch_idx,
+                            y=labels,
+                            n_samples=n_samples,
+                            transform_batch=transform_batch,
+                        )[..., gene_mask]
+                        * scaling
+                    ).cpu()
+                )
+            ]
+
+        if n_samples > 1:
+            # The -2 axis correspond to cells.
+            exprs = np.concatenate(exprs, axis=-2)
+        else:
+            exprs = np.concatenate(exprs, axis=0)
+
+        if n_samples > 1 and return_mean:
+            exprs = exprs.mean(0)
+
+        if return_numpy is None or return_numpy is False:
+            return pd.DataFrame(
+                exprs,
+                columns=adata.var_names[gene_mask],
+                index=adata.obs_names[indices],
+            )
+        else:
+            return exprs
+
+    def differential_expression(
+        self,
+        groupby,
+        group1,
```

Comment:
group1 group2 is not good. Use reference and another word that people generally use. Otherwise it is hard to interpret the LFC sign.